backend/cmm_helpers.ml

```diff
@@ -2336,17 +2336,45 @@ let ptr_offset ptr offset dbg =
 let direct_apply lbl ty args (pos, _mode) dbg =
   Cop (Capply (ty, pos), Cconst_symbol (lbl, dbg) :: args, dbg)
 
+let split_two_halves ~len l =
+  let rec aux len first_halve l =
+    if len = 0
+    then List.rev first_halve, l
+    else
+      match l with
+      | [] -> List.rev first_halve, []
+      | x :: l -> aux (len - 1) (x :: first_halve) l
+  in
+  aux len [] l
+
+let split_arity_for_apply arity args =
+  if List.length arity >= Lambda.max_arity ()
```

Comment:
It's probably slightly better to use `List.compare_length_with` here. Then `really_call_caml_apply` calculation isn't quadratic in the length of the list.